src/main/java/org/logstash/dissect/Dissector.java

```diff
@@ -53,34 +56,58 @@ public Dissector(String mapping) {
         Collections.sort(saveableFields, new FieldComparator());
     }
 
-    public int dissect(byte[] source, Map<String, Object> map) {
-        int pos = dissectValues(source);
-        ValueResolver resolver = new ValueResolver(source, fieldValueRefMap);
+    public int dissect(byte[] source, Map<String, Object> keyValueMap) {
+        ResolveValue resolveValue = resolve(source);
 
         for (Field field : saveableFields) {
-            field.append(map, resolver);
+            field.append(keyValueMap, resolveValue.resolver);
         }
-        return pos;
+        return resolveValue.position;
     }
 
     public int dissect(byte[] source, Event event) {
-        int pos = dissectValues(source);
-        ValueResolver resolver = new ValueResolver(source, fieldValueRefMap);
+        ResolveValue resolveValue = resolve(source);
 
         for (Field field : saveableFields) {
-            field.append(event, resolver);
+            field.append(event, resolveValue.resolver);
         }
-        return pos;
+        return resolveValue.position;
+    }
+
+    private ResolveValue resolve(byte[] source) {
+        final Map<Field, ValueRef> fieldValueRefMap = createFieldValueRefMap();
+        int pos = dissectValues(source, fieldValueRefMap);
+        ValueResolver resolver = new ValueResolver(source, fieldValueRefMap);
+        return new ResolveValue(pos, resolver);
     }
 
-    private int dissectValues(byte[] source) {
-        if (source.length == 0) return 0;
+    private class ResolveValue {
+        private final int position;
+        private final ValueResolver resolver;
+
+        public ResolveValue(int position, ValueResolver resolver) {
+            this.position = position;
+            this.resolver = resolver;
+        }
+    }
+
+    private Map<Field, ValueRef> createFieldValueRefMap() {
+        Map<Field, ValueRef> map = new HashMap<>();
+        map.put(NormalField.getMissing(), new ValueRef(0, 0));
```

Comment:
simplify to

```java
private Map<Field, ValueRef> createFieldValueRefMap() {
    return Collections.singletonMap<>(NormalField.getMissing(), new ValueRef(0, 0));
}
```

But be aware that `Collection.singletonMap()` returns an unmodifiable map (so you can't put any more values in it).

You could also think of making `ValueRef(0, 0)` a constant thus reducing a bit of garbage (provided `ValueRef` will not change).